kernel_npem: add support for kernel NPEM driver

stuarthayes opened the pull request:
Add a new controller for the recently-added kernel NPEM/_DSM driver (in the kernel at drivers/pci/npem.c) that provides access to NPEM LEDs via sysfs. This allows access even if the kernel doesn't allow direct access to PCI config space (for NPEM), and also allows access to the PCI _DSM method of controlling PCIe SSD LEDs (PCI Firmware Specification, Revision 3.3, section 4.7 "_DSM Definitions for PCIe SSD Status LED").

The code was based on the npem controller code, but uses the sysfs interface rather than directly accessing the NPEM capability in PCI config space.

stuartwhayes committed:
kernel_npem: add support for kernel NPEM driver

Add a new controller for the recently-added kernel NPEM/_DSM driver
(in the kernel at drivers/pci/npem.c) that provides access to NPEM LEDs via
sysfs. This allows access even if the kernel doesn't allow direct access to
PCI config space (for NPEM), and also allows access to the PCI _DSM method
of controlling PCIe SSD LEDs (PCI Firmware Specification, Revision 3.3,
section 4.7 "_DSM Definitions for PCIe SSD Status LED").

The code was based on the npem controller code, but uses the sysfs
interface rather than directly accessing the NPEM capability in PCI config
space.

Signed-off-by: Stuart Hayes <stuart.w.hayes@gmail.com>

mtkaczyk reviewed (dismissed):
Very nice, thank you!

bkucman commented on src/lib/block.c:
```diff
@@ -137,6 +141,8 @@ static char *_get_host(char *path, struct cntrl_device *cntrl)
 		result = ahci_get_port_path(path);
+	else if (cntrl->cntrl_type == LED_CNTRL_TYPE_KERNEL_NPEM)
+		result = kernel_npem_get_path(cntrl->sysfs_path);
 	else if (cntrl->cntrl_type == LED_CNTRL_TYPE_DELLSSD)
```
You should add the npem_kernel exclusion to the function is_host_id_supported()

bkucman commented on src/lib/kernel_npem.c:
```diff
@@ -0,0 +1,205 @@
+	struct stat sb;
+	u32 supported;
+
+	supported = 0;
```
```suggestion
	u32 supported = 0;
```

stuarthayes replied:
Fixed

bkucman reviewed (changes requested):
Hi Stuart ,

Thank you for your contribution, Nice work!
After the review, I have a few points to clarify and some minor corrections.

It seems that the NPEM kernel still operates per slots, so my question is: did you use the implemented slot tests for the "NPEM kernel"? If not, please run them and verify that everything works correctly. 

Also, please rebase the branch.

Thanks!

stuartwhayes committed:
kernel_npem: A few cosmetic changes

A few cosmetic changes for kernel_npem.

Signed-off-by: Stuart Hayes <stuart.w.hayes@gmail.com>

stuartwhayes committed:
Update slot support for kernel_npem

Include LED_CNTRL_TYPE_KERNEL_NPEM in led_controller_slot_support().

Signed-off-by: Stuart Hayes <stuart.w.hayes@gmail.com>

mtkaczyk commented on src/lib/block.c:
```diff
@@ -137,6 +141,8 @@ static char *_get_host(char *path, struct cntrl_device *cntrl)
+		result = kernel_npem_get_path(cntrl->sysfs_path);
 	else if (cntrl->cntrl_type == LED_CNTRL_TYPE_DELLSSD)
 		result = dellssd_get_path(cntrl->sysfs_path);
 	else if (cntrl->cntrl_type == LED_CNTRL_TYPE_VMD)
```
Good catch!
Both NPEM and VMD may exists and this is a logic to ensure that NPEM has higher precedence over VMD. Of course it can only happen on setups with VMD turned on.
I think we have to ensure that kernel driver has the highest priority now.

stuartwhayes committed:
Add KERNEL_NPEM controller type in two missed places

Add the newly-added KERNEL_NPEM controller type to two places in block.c
that were missed in the initial commit.

Signed-off-by: Stuart Hayes <stuart.w.hayes@gmail.com>

stuartwhayes committed:
Update tests for KERNEL_NPEM

Update lib_unit_test.c to allow 7 controller types so it doesn't fail with
KERNEL_NPEM.

Signed-off-by: Stuart Hayes <stuart.w.hayes@gmail.com>

stuarthayes commented on src/lib/sysfs.c:
```diff
@@ -413,6 +414,11 @@ static void _scan_slots(struct led_ctx *ctx)
 			if (slot)
 				list_append_ctx(&ctx->sys.slots_list, slot, ctx);
 		}
+		if (cntrl_device->cntrl_type == LED_CNTRL_TYPE_KERNEL_NPEM) {
```
Fixed

stuarthayes commented on src/lib/kernel_npem.h:
```diff
@@ -0,0 +1,53 @@
+status_t kernel_npem_set_state(struct slot_property *slot, enum led_ibpi_pattern state);
+
+/**
+ * @brief Initializes a slot_property for a specified npem controller.
```
Fixed

stuarthayes commented on src/lib/kernel_npem.h:
```diff
@@ -0,0 +1,53 @@
+/**
+ * @brief Initializes a slot_property for a specified npem controller.
+ *
+ * @param[in]         npem_cntrl       Specified npem controller for this slot
```
Fixed

tasleson commented on src/lib/kernel_npem.c:
```diff
@@ -0,0 +1,205 @@
+		make_led_path(led_path, sysfs_path, kernel_npem_leds[i].sysfs_led_name);
+		snprintf(val_text, sizeof(val_text), val & kernel_npem_leds[i].bitmask ? "1" : "0");
+		if (!stat(led_path, &sb))
+			buf_write(led_path, val_text);
```
not checking return code of `buf_write(led_path, val_text);` , but I suppose it failing is remote here.

tasleson commented:
FYI: I'm in the middle of reviewing, testing etc.  I should be done by end of day Tuesday. @stuarthayes Thanks for this PR, greatly appreciated!

mtkaczyk commented:
> How about LED_IBPI_PATTERN_NORMAL ?

For example, for NPEM it means that OK capability is ON. This would be also kind of inconsistently when we are presetting NORMAL when all LEDs are OFF. Unknown at least suggests that we don't know what is happening.

> IMHO we need to let the API caller know that we have some level of hardware support, LED_IBPI_PATTERN_UNKNOWN doesn't convey that.

That would be perfect but probably we don't have such info in all cases. For NPEM, this information is visible as NPEM_ENABLE  and it is turned on persistently when first NPEM request happens. There is no option to disable this and NPEM enable is not presented by driver:
https://elixir.bootlin.com/linux/v6.18/source/drivers/pci/npem.c#L187
https://elixir.bootlin.com/linux/v6.18/source/drivers/pci/npem.c#L207

```c
	/* This bit is always required */
	ctrl = inds | PCI_NPEM_CTRL_ENABLE;
```

So, if all kernel interface NPEM capabilities are off before ledctl is used first time - it is hardware managed but if we turned NORMAL and disabled NORMAL, all are off but it is not hardware managed because NPEM_ENABLE is kept.
NPEM_ENABLE is not presented to userspace.

But overall, If we have good argument for that we can change UNKNOWN to NORMAL, I'm not against. It cannot be handled perfectly because kernel driver interface does not provide functionality to determine it (NPEM_ENABLE).

We have chosen to not present NPEM_ENABLE on purpose because for the regular user it provides unnecessary level of complexity and was not kernel ledclass interface friendly.

mtkaczyk commented:
@stuarthayes gentle reminder :)

stuartwhayes committed:
Return LOCATE_OFF when only locate state supported

Many systems (Dell servers) support the LED control _DSM but only support
the locate state. Return LOCATE_OFF instead of UNKNOWN for these sysetms
when LOCATE isn't on.

Signed-off-by: Stuart Hayes <stuart.w.hayes@gmail.com>

stuartwhayes committed:
Only support one NPEM controller

Change the new "Kernel NPEM" controller to be the "NPEM" controller,
replacing the older NPEM controller (that supports NPEM via direct PCI
config space access) by default. Add a conf file option "USERSPACE_NPEM="
that can be specified to force use of the older NPEM driver, for use on
systems with kernels that don't have the kernel NPEM/_DSM driver included.

Signed-off-by: Stuart Hayes <stuart.w.hayes@gmail.com>

mtkaczyk commented:
Moved to #287
